sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -182,6 +182,10 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   /** Specifies sort order for each partition requirements on the input data for this operator. */
   def requiredChildOrdering: Seq[Seq[SortOrder]] = Seq.fill(children.size)(Nil)
 
+  private val executeRDD = LazyTry {
```

Comment:
ideally we only need to cache it for the root plan node?

Reply:
Currently it's already cached at shuffle boundaries https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala#L257
and we do want that for when the plans gets reused or also for when fragments of plan are moved around by AQE.

We don't really need to cache it for plans that are not roots of some execution pipeline (Stage).
I get your worry of if this would make us needlessly hold references to some throwaway intermediate RDDs that are just accidentally triggered... Hmm...
For anything that is then actually used in the Stages, references to it would be kept by the pipeline roots anyway... The question is whether there are some accidental calls to `execute` that will now keep these references even though it's not used at all... E.g. an `execute` call to a non-WSCG operator inside WSCG, creating and saving the non-WSCG RDD, while actual execution uses WSCG...
I don't think so, though such things happen via accident / non-intended side effect.
I think it's a bigger issue that Dataset holding references to QueryExecution, holding references to SparkPlans, holding references to other stuff results in a lot of references that impedes garbage collection.
E.g.  https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala#L257 leads to shuffle data not being cleaned up from disk while the user's program holds a live reference to Dataset.

Running in Spark Connect mode actually makes a lot of such issues go away, as all objects created during connect execution have lifetime that's finished after connect run finishes.
Here I'm trying to solve a different problem, where it's inconsistent what objects are reused vs. recreated from scratch, resulting in duplication of side effects.